datafusion/physical-expr/src/expressions/case/literal_lookup_table/primitive_lookup_table.rs

```diff
@@ -0,0 +1,229 @@
+    T::Native: ToHashableKey,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PrimitiveIndexMap")
+            .field("map", &self.map)
+            .finish()
+    }
+}
+
+impl<T> PrimitiveIndexMap<T>
+where
+    T: ArrowPrimitiveType,
+    T::Native: ToHashableKey,
+{
+    /// Try creating a new lookup table from the given literals and else index.
+    /// The index of each literal in the vector is used as the mapped value in the lookup table.
+    ///
+    /// `literals` are guaranteed to be unique and non-nullable
+    pub(super) fn try_new(
+        unique_non_null_literals: Vec<ScalarValue>,
+    ) -> datafusion_common::Result<Self> {
+        let input = ScalarValue::iter_to_array(unique_non_null_literals)?;
+
+        // Literals are guaranteed to not contain nulls
+        if input.null_count() > 0 {
+            return internal_err!("Literal values for WHEN clauses cannot contain nulls");
+        }
+
+        let map = input
+            .as_primitive::<T>()
+            .values()
+            .iter()
+            .enumerate()
+            // Because literals are unique we can collect directly, and we can avoid only inserting the first occurrence
+            .map(|(map_index, value)| (value.into_hashable_key(), map_index as u32))
+            .collect();
+
+        Ok(Self {
+            map,
+            data_type: input.data_type().clone(),
+        })
+    }
+
+    fn map_primitive_array_to_when_indices(
+        &self,
+        array: &PrimitiveArray<T>,
+        else_index: u32,
+    ) -> datafusion_common::Result<Vec<u32>> {
+        let indices = array
+            .into_iter()
+            .map(|value| match value {
+                Some(value) => self
+                    .map
+                    .get(&value.into_hashable_key())
+                    .copied()
+                    .unwrap_or(else_index),
+
+                None => else_index,
+            })
+            .collect::<Vec<u32>>();
+
+        Ok(indices)
+    }
+}
+
+impl<T> WhenLiteralIndexMap for PrimitiveIndexMap<T>
+where
+    T: ArrowPrimitiveType,
+    T::Native: ToHashableKey,
+{
+    fn map_to_when_indices(
+        &self,
+        array: &ArrayRef,
+        else_index: u32,
+    ) -> datafusion_common::Result<Vec<u32>> {
+        match array.data_type() {
+            dt if dt == &self.data_type => {
+                let primitive_array = array.as_primitive::<T>();
+
+                self.map_primitive_array_to_when_indices(primitive_array, else_index)
+            }
+            // We support dictionary primitive array as we create the lookup table in `CaseWhen` expression
+            // creation when we don't know the schema, so we may receive dictionary encoded primitive arrays at execution time.
+            DataType::Dictionary(_, value_type)
+                if value_type.as_ref() == &self.data_type =>
+            {
+                // Cast here to simplify the implementation.
```

Comment:
as above, you can probably reduce the size of this code by always calling `cast`